src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -173,10 +185,26 @@ dfsForest g = dfsForestFrom (vertexList g) g
 --  'forest' (dfsForestFrom [3] $ 'circuit' [1..5] + 'circuit' [5,4..1]) == 'path' [3,2,1,5,4]
 -- @
 dfsForestFrom :: Ord a => [a] -> AdjacencyMap a -> Forest a
-dfsForestFrom vs = Typed.dfsForestFrom vs . Typed.fromAdjacencyMap
+dfsForestFrom vs g = dfsForestFrom' [ v | v <- vs, hasVertex v g ] g
+
+dfsForestFrom' :: Ord a => [a] -> AdjacencyMap a -> Forest a
+dfsForestFrom' vs g = evalState (explore vs) mempty where
+  explore (v:vs) = discovered v >>= \case
+    True -> (:) <$> walk v <*> explore vs
+    False -> explore vs 
+  explore [] = return []
+  walk v = Node v <$> explore (adjacent v)
+  adjacent v = Set.toList (postSet v g)
```

Comment:
I find it interesting that `postSet` is the only function on graphs that you need. This means you could probably write it generically, for any `ToGraph`, and simply re-export it here with a nicer specialised type (unless there are some nasty import cycles).

Reply:
On second thought, it's not clear if a generic implementation is really useful: it may be faster to convert a whole graph to an `AdjacencyMap` instead of computing `postSet` on a vertex-by-vertex basis.